File: openvidu-call-back/src/services/OpenViduService.ts
import { Connection, ConnectionProperties, OpenVidu, OpenViduRole, Recording, Session, SessionProperties } from 'openvidu-node-client';
import { OPENVIDU_SECRET, OPENVIDU_URL } from '../config';
import { RetryOptions } from '../utils';

export class OpenViduService {
	MODERATOR_TOKEN_NAME = 'ovCallModeratorToken';
	PARTICIPANT_TOKEN_NAME = 'ovCallParticipantToken';
	moderatorsCookieMap: Map<string, { token: string; recordingId: string }> = new Map<string, { token: string; recordingId: string }>();
	participantsCookieMap: Map<string, string[]> = new Map<string, string[]>();

	protected static instance: OpenViduService;
	private openvidu: OpenVidu;
	private edition: string;

	private constructor() {
		this.openvidu = new OpenVidu(OPENVIDU_URL, OPENVIDU_SECRET);
		if (process.env.NODE_ENV === 'production') this.openvidu.enableProdMode();
	}

	static getInstance() {
		if (!OpenViduService.instance) {
			OpenViduService.instance = new OpenViduService();
		}
		return OpenViduService.instance;
	}

	getBasicAuth(): string {
		return this.openvidu.basicAuth;
	}

	getDateFromCookie(cookies: any): number {
		try {
			const cookieToken = cookies[this.MODERATOR_TOKEN_NAME] || cookies[this.PARTICIPANT_TOKEN_NAME];
			if (!!cookieToken) {
				const cookieTokenUrl = new URL(cookieToken);
				const date = cookieTokenUrl?.searchParams.get('createdAt');
				return Number(date);
			} else {
				return Date.now();
			}
		} catch (error) {
			return Date.now();
		}
	}

	getSessionIdFromCookie(cookies: any): string {
		try {
			const cookieToken = cookies[this.MODERATOR_TOKEN_NAME] || cookies[this.PARTICIPANT_TOKEN_NAME];
			const cookieTokenUrl = new URL(cookieToken);
			return cookieTokenUrl?.searchParams.get('sessionId');
		} catch (error) {
			console.log('Session cookie not found', cookies);
			console.error(error);
			return '';
		}
	}

	getSessionIdFromRecordingId(recordingId: string): string {
		return recordingId.split('~')[0];
	}

	isModeratorSessionValid(sessionId: string, cookies: any): boolean {
		try {
			if (!this.moderatorsCookieMap.has(sessionId)) return false;
			if (!cookies[this.MODERATOR_TOKEN_NAME]) return false;
			const storedTokenUrl = new URL(this.moderatorsCookieMap.get(sessionId).token);
			const cookieTokenUrl = new URL(cookies[this.MODERATOR_TOKEN_NAME]);
			const cookieSessionId = cookieTokenUrl.searchParams.get('sessionId');
			const cookieToken = cookieTokenUrl.searchParams.get(this.MODERATOR_TOKEN_NAME);
			const cookieDate = cookieTokenUrl.searchParams.get('createdAt');

			const storedToken = storedTokenUrl.searchParams.get(this.MODERATOR_TOKEN_NAME);
			const storedDate = storedTokenUrl.searchParams.get('createdAt');

			return sessionId === cookieSessionId && cookieToken === storedToken && cookieDate === storedDate;
		} catch (error) {
			return false;
		}
	}

	isParticipantSessionValid(sessionId: string, cookies: any): boolean {
		try {
			if (!this.participantsCookieMap.has(sessionId)) return false;
			if (!cookies[this.PARTICIPANT_TOKEN_NAME]) return false;
			const storedTokens: string[] | undefined = this.participantsCookieMap.get(sessionId);
			const cookieTokenUrl = new URL(cookies[this.PARTICIPANT_TOKEN_NAME]);
			const cookieSessionId = cookieTokenUrl.searchParams.get('sessionId');
			const cookieToken = cookieTokenUrl.searchParams.get(this.PARTICIPANT_TOKEN_NAME);
			const cookieDate = cookieTokenUrl.searchParams.get('createdAt');

			return (
				storedTokens?.some((token) => {
					const storedTokenUrl = new URL(token);
					const storedToken = storedTokenUrl.searchParams.get(this.PARTICIPANT_TOKEN_NAME);
					const storedDate = storedTokenUrl.searchParams.get('createdAt');
					return sessionId === cookieSessionId && cookieToken === storedToken && cookieDate === storedDate;
				}) ?? false
			);
		} catch (error) {
			return false;
		}
	}

	public async createSession(sessionId: string, retryOptions = new RetryOptions()): Promise<Session> {
		while (retryOptions.canRetry()) {
			try {
				console.log('Creating session: ', sessionId);
				let sessionProperties: SessionProperties = { customSessionId: sessionId };
				const session = await this.openvidu.createSession(sessionProperties);
				await session.fetch();
				return session;
			} catch (error) {
				const status = error.message;
				if ((status >= 500 && status <= 504) || status == 404) {
					// Retry is used for OpenVidu Enterprise High Availability for reconnecting purposes
					// to allow fault tolerance
					// 502 to 504 are returned when OpenVidu Server is not available (stopped, not reachable, etc...)
					// 404 is returned when the session does not exist which is returned by fetch operation in createSession
					// and it is not a possible error after session creation
					console.log('Error creating session: ', status, 'Retrying session creation...', retryOptions);
					await retryOptions.retrySleep();
				} else {
					console.log("Unknown error creating session: ", error);
					throw error;
				}
			}
		}
		throw new Error('Max retries exceeded while creating connection');
	}

	public async createConnection(session: Session, nickname: string, role: OpenViduRole, retryOptions = new RetryOptions()): Promise<Connection> {
		while (retryOptions.canRetry()) {
			try {
				console.log(`Requesting token for session ${session.sessionId}`);
				let connectionProperties: ConnectionProperties = { role };
				if (!!nickname) {
					connectionProperties.data = JSON.stringify({
						openviduCustomConnectionId: nickname
					});
				}
				console.log('Connection Properties:', connectionProperties);
				const connection = await session.createConnection(connectionProperties);
				this.edition = new URL(connection.token).searchParams.get('edition');
				return connection;
			} catch (error) {
				const status = Number(error.message);
				if (status >= 500 && status <= 504) {
					// Retry is used for OpenVidu Enterprise High Availability for reconnecting purposes
					// to allow fault tolerance
					console.log('Error creating connection: ', status, 'Retrying connection creation...', retryOptions);
					await retryOptions.retrySleep();
				} else {
					console.log("Unknown error creating connection: ", error);
					throw error;
				}
			}
		}
		throw new Error('Max retries exceeded while creating connection');
	}

	public async startRecording(sessionId: string): Promise<Recording> {
		return this.openvidu.startRecording(sessionId);
	}

	public stopRecording(recordingId: string): Promise<Recording> {
		return this.openvidu.stopRecording(recordingId);
	}

	public deleteRecording(recordingId: string): Promise<Error> {
		return this.openvidu.deleteRecording(recordingId);
	}
	public getRecording(recordingId: string): Promise<Recording> {
		return this.openvidu.getRecording(recordingId);
	}

	public async listAllRecordings(): Promise<Recording[]> {
		return await this.openvidu.listRecordings();
	}

	public async listRecordingsBySessionIdAndDate(sessionId: string, date: number) {
		const recordingList: Recording[] = await this.listAllRecordings();
		return recordingList.filter((recording) => {
			const recordingDateEnd = recording.createdAt + recording.duration * 1000;
			return recording.sessionId === sessionId && recordingDateEnd >= date;
		});
	}

	public async startBroadcasting(sessionId: string, url: string): Promise<void> {
		return this.openvidu.startBroadcast(sessionId, url);
	}

	public async stopBroadcasting(sessionId: string): Promise<void> {
		return this.openvidu.stopBroadcast(sessionId);
	}

	public isPRO(): boolean {
		return this.edition.toUpperCase() === 'PRO';
	}

	public isCE(): boolean {
		return this.edition.toUpperCase() === 'CE';
	}
}

File: openvidu-call-back-java/src/main/java/io/openvidu/call/java/services/OpenViduService.java
package io.openvidu.call.java.services;

import java.util.ArrayList;
import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.annotation.PostConstruct;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.util.MultiValueMap;
import org.springframework.web.util.UriComponentsBuilder;

import io.openvidu.call.java.models.RecordingData;
import io.openvidu.call.java.utils.RetryException;
import io.openvidu.call.java.utils.RetryOptions;
import io.openvidu.java.client.Connection;
import io.openvidu.java.client.ConnectionProperties;
import io.openvidu.java.client.OpenVidu;
import io.openvidu.java.client.OpenViduHttpException;
import io.openvidu.java.client.OpenViduJavaClientException;
import io.openvidu.java.client.OpenViduRole;
import io.openvidu.java.client.Recording;
import io.openvidu.java.client.Session;
import io.openvidu.java.client.SessionProperties;

@Service
public class OpenViduService {

	public static final String MODERATOR_TOKEN_NAME = "ovCallModeratorToken";
	public static final String PARTICIPANT_TOKEN_NAME = "ovCallParticipantToken";
	public Map<String, RecordingData> moderatorsCookieMap = new HashMap<String, RecordingData>();
	public Map<String, List<String>> participantsCookieMap = new HashMap<String, List<String>>();

	@Value("${OPENVIDU_URL}")
	public String OPENVIDU_URL;

	@Value("${OPENVIDU_SECRET}")
	private String OPENVIDU_SECRET;

	private OpenVidu openvidu;
	private String edition;

	@PostConstruct
	public void init() {
		this.openvidu = new OpenVidu(OPENVIDU_URL, OPENVIDU_SECRET);
	}

	public String getBasicAuth() {
		String stringToEncode = "OPENVIDUAPP:" + OPENVIDU_SECRET;
		String encodedString = Base64.getEncoder().encodeToString(stringToEncode.getBytes());
		return "Basic " + new String(encodedString);
	}

	public boolean isPRO() {
		return this.edition.toUpperCase().equals("PRO");
	}

	public boolean isCE() {
		return this.edition.toUpperCase().equals("CE");
	}

	public long getDateFromCookie(String recordingToken) {
		try {
			if (!recordingToken.isEmpty()) {
				MultiValueMap<String, String> cookieTokenParams = UriComponentsBuilder.fromUriString(recordingToken)
						.build()
						.getQueryParams();
				String date = cookieTokenParams.get("createdAt").get(0);
				return Long.parseLong(date);
			} else {
				return System.currentTimeMillis();
			}
		} catch (Exception e) {
			return System.currentTimeMillis();
		}
	}

	public String getSessionIdFromCookie(String cookie) {
		try {

			if (!cookie.isEmpty()) {
				MultiValueMap<String, String> cookieTokenParams = UriComponentsBuilder.fromUriString(cookie)
						.build().getQueryParams();
				return cookieTokenParams.get("sessionId").get(0);
			}

		} catch (Exception error) {
			System.out.println("Session cookie not found");
			System.err.println(error);
		}
		return "";

	}

	public String getSessionIdFromRecordingId(String recordingId) {
		return recordingId.split("~")[0];
	}

	public boolean isModeratorSessionValid(String sessionId, String token) {
		try {

			if(token.isEmpty()) return false;
			if(!this.moderatorsCookieMap.containsKey(sessionId)) return false;

			MultiValueMap<String, String> storedTokenParams = UriComponentsBuilder
					.fromUriString(this.moderatorsCookieMap.get(sessionId).getToken()).build().getQueryParams();

			MultiValueMap<String, String> cookieTokenParams = UriComponentsBuilder
					.fromUriString(token).build().getQueryParams();

			String cookieSessionId = cookieTokenParams.get("sessionId").get(0);
			String cookieToken = cookieTokenParams.get(MODERATOR_TOKEN_NAME).get(0);
			String cookieDate = cookieTokenParams.get("createdAt").get(0);

			String storedToken = storedTokenParams.get(MODERATOR_TOKEN_NAME).get(0);
			String storedDate = storedTokenParams.get("createdAt").get(0);

			return sessionId.equals(cookieSessionId) && cookieToken.equals(storedToken)
					&& cookieDate.equals(storedDate);

		} catch (Exception e) {
			return false;
		}
	}

	public boolean isParticipantSessionValid(String sessionId, String cookie) {

		try {
			if (!this.participantsCookieMap.containsKey(sessionId))	return false;
			if(cookie.isEmpty()) return false;


			MultiValueMap<String, String> cookieTokenParams = UriComponentsBuilder
					.fromUriString(cookie).build().getQueryParams();

			List<String> storedTokens = this.participantsCookieMap.get(sessionId);

			String cookieSessionId = cookieTokenParams.get("sessionId").get(0);
			String cookieToken = cookieTokenParams.get(PARTICIPANT_TOKEN_NAME).get(0);
			String cookieDate = cookieTokenParams.get("createdAt").get(0);

			for (String token : storedTokens) {
				MultiValueMap<String, String> storedTokenParams = UriComponentsBuilder
					.fromUriString(token).build().getQueryParams();

				String storedToken = storedTokenParams.get(PARTICIPANT_TOKEN_NAME).get(0);
				String storedDate = storedTokenParams.get("createdAt").get(0);

				if (sessionId.equals(cookieSessionId) && cookieToken.equals(storedToken) && cookieDate.equals(storedDate)) {
					return true;
				}
			}

			return false;

		} catch (Exception e) {
			return false;
		}
	}

	public Session createSession(String sessionId)
			throws OpenViduJavaClientException, OpenViduHttpException, InterruptedException, RetryException {
		RetryOptions retryOptions = new RetryOptions();
		return createSession(sessionId, retryOptions);
	}

	private Session createSession(String sessionId, RetryOptions retryOptions)
			throws OpenViduJavaClientException, OpenViduHttpException, InterruptedException, RetryException {
		while(retryOptions.canRetry()) {
			try {
				Map<String, Object> params = new HashMap<String, Object>();
				params.put("customSessionId", sessionId);
				SessionProperties properties = SessionProperties.fromJson(params).build();
				Session session = openvidu.createSession(properties);
				session.fetch();
				return session;
			} catch (OpenViduHttpException e) {
				if ((e.getStatus() >= 500 && e.getStatus() <= 504) || e.getStatus() == 404) {
					// Retry is used for OpenVidu Enterprise High Availability for reconnecting purposes
					// to allow fault tolerance
					// 502 to 504 are returned when OpenVidu Server is not available (stopped, not reachable, etc...)
					// 404 is returned when the session does not exist which is returned by fetch operation in createSession
					// and it is not a possible error after session creation
					System.err.println("Error creating session: " + e.getMessage()
						+ ". Retrying session creation..." + retryOptions.toString());
					retryOptions.retrySleep();
				} else {
					System.err.println("Error creating session: " + e.getMessage());
					throw e;
				}
			}
		}
		throw new RetryException("Max retries exceeded");
	}

	public Connection createConnection(Session session, String nickname, OpenViduRole role)
			throws OpenViduJavaClientException, OpenViduHttpException, RetryException, InterruptedException {
		return createConnection(session, nickname, role, new RetryOptions());
	}

	private Connection createConnection(Session session, String nickname, OpenViduRole role, RetryOptions retryOptions)
			throws OpenViduJavaClientException, OpenViduHttpException, RetryException, InterruptedException {
		Map<String, Object> params = new HashMap<String, Object>();
		Map<String, Object> connectionData = new HashMap<String, Object>();

		if (!nickname.isEmpty()) {
			connectionData.put("openviduCustomConnectionId", nickname);
		}
		params.put("role", role.name());
		params.put("data", connectionData.toString());
		ConnectionProperties properties = ConnectionProperties.fromJson(params).build();

		Connection connection = null;
		while (retryOptions.canRetry()) {
			try {
				connection = session.createConnection(properties);
				break;
			} catch (OpenViduHttpException e) {
				if (e.getStatus() >= 500 && e.getStatus() <= 504) {
					// Retry is used for OpenVidu Enterprise High Availability for reconnecting purposes
					// to allow fault tolerance
					System.err.println("Error creating connection: " + e.getMessage()
						+ ". Retrying connection creation..." + retryOptions.toString());
					retryOptions.retrySleep();
				} else {
					System.err.println("Error creating connection: " + e.getMessage());
					throw e;
				}
			}
		}

		if (connection == null) {
			throw new RetryException("Max retries exceeded");
		}

		MultiValueMap<String, String> tokenParams = UriComponentsBuilder
				.fromUriString(connection.getToken()).build().getQueryParams();

		if (tokenParams.containsKey("edition")) {
			this.edition = tokenParams.get("edition").get(0);
		} else {
			this.edition = "ce";
		}

		return connection;

	}

	public Recording startRecording(String sessionId) throws OpenViduJavaClientException, OpenViduHttpException {
		return this.openvidu.startRecording(sessionId);
	}

	public Recording stopRecording(String recordingId) throws OpenViduJavaClientException, OpenViduHttpException {
		return this.openvidu.stopRecording(recordingId);
	}

	public void deleteRecording(String recordingId) throws OpenViduJavaClientException, OpenViduHttpException {
		this.openvidu.deleteRecording(recordingId);
	}

	public Recording getRecording(String recordingId) throws OpenViduJavaClientException, OpenViduHttpException {
		return this.openvidu.getRecording(recordingId);
	}

	public List<Recording> listAllRecordings() throws OpenViduJavaClientException, OpenViduHttpException {
		return this.openvidu.listRecordings();
	}

	public List<Recording> listRecordingsBySessionIdAndDate(String sessionId, long date)
			throws OpenViduJavaClientException, OpenViduHttpException {
		List<Recording> recordings = this.listAllRecordings();
		List<Recording> recordingsAux = new ArrayList<Recording>();
		for (Recording recording : recordings) {
			if (recording.getSessionId().equals(sessionId) && recording.getCreatedAt() + recording.getDuration() * 1000 >= date) {
				recordingsAux.add(recording);
			}
		}
		return recordingsAux;
	}

	public void startBroadcast(String sessionId, String broadcastUrl)
			throws OpenViduJavaClientException, OpenViduHttpException {
		this.openvidu.startBroadcast(sessionId, broadcastUrl);
	}

	public void stopBroadcast(String sessionId) throws OpenViduJavaClientException, OpenViduHttpException {
		this.openvidu.stopBroadcast(sessionId);
	}

}
